src/agent/onefuzz-supervisor/src/auth.rs

```diff
@@ -147,20 +167,32 @@ impl From<ClientAccessTokenBody> for AccessToken {
 #[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
 pub struct ManagedIdentityCredentials {
     resource: String,
+    multi_tenant_domain: Option<String>,
 }
 
 const MANAGED_IDENTITY_URL: &str =
     "http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01";
 
 impl ManagedIdentityCredentials {
-    pub fn new(resource: String) -> Self {
-        Self { resource }
+    pub fn new(resource: String, multi_tenant_domain: Option<String>) -> Self {
+        Self {
+            resource,
+            multi_tenant_domain,
+        }
     }
 
     fn url(&self) -> Url {
         let mut url = Url::parse(MANAGED_IDENTITY_URL).unwrap();
-        url.query_pairs_mut()
-            .append_pair("resource", &self.resource);
+        let resource = if let Some(domain) = &self.multi_tenant_domain {
+            let uri = Url::parse(&self.resource).unwrap();
```

Comment:
Just a suggestion, as pointed out before, this `unwrap` may panic if it's an invalid url. Since the return type is not `Result<>`,  an alternative could be using [expect](https://learning-rust.github.io/docs/e4.unwrap_and_expect.html#expect), which will still panic but allows you to pass a string message.

Reply:
That's a better alternative, but still panics, which we want to avoid.

I suggest doing some validation earlier. For example, you could validate and precompute `multi_tenant_domain` in the ctor (`ManagedIdentityCredentials::new()`), which would then return a `Result`.

@anslutsk, using some stand-in values, could you share what we expect `resource` and a valid `managed_identity_url` to look like?

Reply:
I put in some temporary println statements to provide in context here.

In the case where OneFuzz is deployed in single tenant mode, the config.json file contains an entry for "multi_tenant_config" which has value "null" and the self.resource String value is "https://anslutsk-26.azurewebsites.net" and thus, the pattern "if let Some(domain)" is not triggered, so the else clause is run, which executes the line: self.resource.clone().

In the case where OneFuzz is deployed in multi tenant mode, the config.json file contains the entry "multi_tenant_config" with the value "mspmecloud.onmicrosoft.com" and the self.resource String value is "https://anslutsk-26.azurewebsites.net" and so the pattern "if let Some(domain)" is triggered, so the corresponding code block is executed and the value of self.instance[0] is "anslutsk-26"

Reply:
When there was an invalid domain in the "multi_tenant_config" json config item, for example the String value "**....**" the program did not crash and we get the expected error during while trying to obtain a Bearer token from IMS:

[2021-02-24T07:12:19Z ERROR onefuzz_telemetry] error running supervisor agent: 400 Bad Request: {"error":"invalid_resource","error_description":"AADSTS500011: The resource principal named https://**....**/anslutsk-multi-26 was not found in the tenant named 975f013f-7f24-47e8-a7d3-abc4752bf346. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You might have sent your authentication request to the wrong tenant.\r\nTrace ID: 0eec3338-bfa4-49f1-a7aa-7d13f9842200\r\nCorrelation ID: 02d18935-c309-4c71-8757-7a9770a98805\r\nTimestamp: 2021-02-24 07:12:19Z","error_codes":[500011],"timestamp":"2021-02-24 07:12:19Z","trace_id":"0eec3338-bfa4-49f1-a7aa-7d13f9842200","correlation_id":"02d18935-c309-4c71-8757-7a9770a98805","error_uri":"https://westus2.login.microsoft.com/error?code=500011"}

In addition:

When there was an invalid domain in the "multi_tenant_config" json config item, for example the String value "**asdfasdfasdf**" the program did not crash and we get the expected error during while trying to obtain a Bearer token from IMS:

[2021-02-24T07:16:33Z ERROR onefuzz_telemetry] error running supervisor agent: 400 Bad Request: {"error":"invalid_resource","error_description":"AADSTS500011: The resource principal named https://**asdfasdfasdf**/anslutsk-multi-26 was not found in the tenant named 975f013f-7f24-47e8-a7d3-abc4752bf346. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You might have sent your authentication request to the wrong tenant.\r\nTrace ID: b0c8b83c-e552-4f2f-9f77-d280785d2600\r\nCorrelation ID: 1982fe67-c50a-4e0a-ac63-186134839aae\r\nTimestamp: 2021-02-24 07:16:33Z","error_codes":[500011],"timestamp":"2021-02-24 07:16:33Z","trace_id":"b0c8b83c-e552-4f2f-9f77-d280785d2600","correlation_id":"1982fe67-c50a-4e0a-ac63-186134839aae","error_uri":"https://westus2.login.microsoft.com/error?code=500011"}

I do not see the value in adding additional error handling. Am I missing something?

Reply:
The value assigned to self.resource is taken from the config item "onefuzz_url" and if the value is invalid, the program crashes like this:

[2021-02-24T15:54:35Z ERROR onefuzz_telemetry] error loading supervisor agent config: invalid value: string "asdf", expected description() is deprecated; use Display at line 1 column 50

If the "onefuzz_url" is set to "null" the following error is displayed:

[2021-02-24T16:03:23Z ERROR onefuzz_telemetry] error loading supervisor agent config: invalid type: null, expected a string representing an URL at line 1 column 48
Error: invalid type: null, expected a string representing an URL at line 1 column 48

Reply:
I don't think it's necessary to change this section of the code, but that doesn't mean that I don't support making a change! If either of you would like to suggest a specific change then I'd be happy to test it. It would need to be pretty specific though since I don't share either of your view points.

Reply:
Thanks for sharing examples of expected/valid inputs, and your investigation into the current failure modes.

> I do not see the point in adding additional error handling. Am I missing something?

We still want to change what we're doing here, because this de facto current unreachability is very non-local, and depends on the behavior of the larger system. But if we ever accidentally introduce a bug elsewhere that makes the `unwrap()` reachable, we want to fail in a way that will be easier to understand and fix, and get [reported here](https://github.com/microsoft/onefuzz/blob/main/src/agent/onefuzz-supervisor/src/main.rs#L109-L113).

So, the constraint here is that we must not `unwrap()`. We could make `ManagedIdentityCredentials::url()` return a `Result<Url>`, but that would be a bit silly, since we'd then have to change all use sites, and a non-`Ok()` result is actually an error that we can detect much earlier (in the constructor).

Instead, I suggest changing `ManagedIdentityCredentials::new()` to have the signature `new(Url, Option<Url>) -> Result<Self>`, taking care to generate the `resource` string _without_ a trailing `/` in the non-multi-tenant case.

Reply:
test123